### PR TITLE
[WIP] Fix Windows 8, 10 cursors drawing as black

### PR DESCRIPTION
**Pull request by perkerk**
Fixes 2 issues:
- xor buffer was misaligned because scanline size was calculated as width * GetBitsPerPixel
- CreateBitmap did not create a usable bitmap for the hbmColor field

### CODE REVIEW

File: client/Windows/wf_graphics.c [hunk 1]
```diff
@@ -237,24 +237,35 @@ static BOOL wf_Pointer_New(rdpContext* context, const rdpPointer* pointer)
...
 		info.hbmMask = CreateBitmap(pointer->width, pointer->height, 1, 1, pdata);
 		_aligned_free(pdata);
-		pdata = (BYTE*) _aligned_malloc(pointer->width * pointer->height *
-		                                GetBitsPerPixel(gdi->dstFormat), 16);
 
-		if (!pdata)
+		UINT32 srcFormat;
+		switch (pointer->xorBpp)
+		{
+		case 16:
```
Comment (akallabeth) [inline client/Windows/wf_graphics.c]: Is this always 15bpp?
Case for 8bpp and 15bpp is missing or are they not supported at all?

**Comment by akallabeth** [comment]
@freerdp-bot test

**Comment by akallabeth** [comment]
Thank you for the contribution, the windows client does really need some love :+1:

File: client/Windows/wf_graphics.c [hunk 2]
```diff
@@ -237,24 +237,35 @@ static BOOL wf_Pointer_New(rdpContext* context, const rdpPointer* pointer)
 		            pointer->height);
 		info.hbmMask = CreateBitmap(pointer->width, pointer->height, 1, 1, pdata);
 		_aligned_free(pdata);
-		pdata = (BYTE*) _aligned_malloc(pointer->width * pointer->height *
-		                                GetBitsPerPixel(gdi->dstFormat), 16);
 
-		if (!pdata)
+		UINT32 srcFormat;
```
Comment (akallabeth) [inline client/Windows/wf_graphics.c]: Declarations at beginning of a scope only, older compilers (hello VS2010) are not really C99 conformant...

File: client/Windows/wf_graphics.c [hunk 3]
```diff
@@ -237,24 +237,35 @@ static BOOL wf_Pointer_New(rdpContext* context, const rdpPointer* pointer)
...
 
 		if (!freerdp_image_copy_from_pointer_data(pdata, gdi->dstFormat, 0, 0, 0,
-		        pointer->width, pointer->height,
-		        pointer->xorMaskData, pointer->lengthXorMask,
-		        pointer->andMaskData, pointer->lengthAndMask, pointer->xorBpp, &gdi->palette))
+			pointer->width, pointer->height,
+			pointer->xorMaskData, pointer->lengthXorMask,
+			pointer->andMaskData, pointer->lengthAndMask, pointer->xorBpp, &gdi->palette))
 		{
-			_aligned_free(pdata);
```
Comment (akallabeth) [inline client/Windows/wf_graphics.c]: possible memory leak in error case here

File: client/Windows/wf_graphics.c (hunk 1 above)
Comment (perkerk) [reply client/Windows/wf_graphics.c]: Well 8bpp was falling through to palette format below. This switch is just a reverse map of what's in the freerdp_image_copy_from_pointer_data() routine, and other resolutions aren't supported by it. I should be more specific in this case statement. 

But, can I just get rid of this case statement and using  freerdp_image_copy_from_pointer_data altogether? Do you know why it's even in there. The and and xor data should already be in a usable format for CreateBitmap and CreateDIBSection (other than flipping it to a top-down bitmap). I should be able to call flip_bitmap on the xor data and pass it directly to CreateDIBSection instead

File: client/Windows/wf_graphics.c (hunk 2 above)
Comment (perkerk) [reply client/Windows/wf_graphics.c]: got it

File: client/Windows/wf_graphics.c (hunk 1 above)
Comment (perkerk) [reply client/Windows/wf_graphics.c]: I will look into what bit depths are supported for color cursors, maybe some <24bpp formats do need to be converted to an RGB (24 or 32) format. I'll also check what happens in the 16bpp case

Comment (akallabeth) [reply client/Windows/wf_graphics.c]: @perkerk You're probably right, the `freerdp_image_copy_from_pointer_data` is used on non windows platforms to convert to usable bitmap formats.

**Comment by perkerk** [comment]
@akallabeth Won't be able to get to this until sometime next week

**Comment by akallabeth** [comment]
@perkerk any update?

**Comment by perkerk** [comment]
@akallabeth fixed in my fork but I have to merge up, will try to get to this over the weekend

**Pull request closed without merging**